lib/common/common/src/mmap_hashmap.rs

```diff
@@ -0,0 +1,233 @@
+use std::collections::{BTreeMap, BTreeSet};
+use std::fs::File;
+use std::io::{self, Cursor, Write as _};
+use std::mem::size_of;
+use std::path::Path;
+use std::str;
+
+use memmap2::Mmap;
+use ph::fmph::Function;
+#[cfg(test)]
+use rand::rngs::StdRng;
+#[cfg(test)]
+use rand::{Rng as _, SeedableRng as _};
+use zerocopy::{AsBytes, FromBytes, FromZeroes};
+
+/// On-disk hash map baked by a memory-mapped file.
+pub struct MmapHashMap {
+    mmap: Mmap,
+    header: Header,
+    phf: Function,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug, AsBytes, FromBytes, FromZeroes)]
+struct Header {
+    buckets_pos: u64,
+    buckets_count: u64,
+}
+
+const PADDING_SIZE: usize = 4096;
+
+impl MmapHashMap {
+    /// Save `map` contents to `path`.
+    pub fn create(path: &Path, map: BTreeMap<String, BTreeSet<u32>>) -> io::Result<()> {
```

Comment:
Can we implement hashmap generic over key? Map field index is defined for `String` and for `IntPayloadType`

Reply:
Yep, I have a plan to make it generic and store
1. For strings, `size: u64` + `contents: [u8; size]`. (implemented in this PR)
2. For ints and other fixed-size values, just `data: T`.

But haven't decided a best way to implement it (either by introducing a trait, or maybe using some existing crate). I want to postpone it for now.

Reply:
Okay, for me it's fine to do it later and make this PR simpler, resolved

Reply:
#4648